text/0000-sandbox-environment.md

```diff
@@ -0,0 +1,276 @@
+
+Rust supports the `env!` and `option_env!` macros which allow Rust programs to
+query arbitrary process environment variables at compilation time. This is a
+very flexible mechanism to pass compile-time information to the program; in
+effect it uses the environment as a form of compile-time directive.
+
+However, in many cases it is too flexible. It poses several problems:
+1. Environment variables are generally not tracked by build systems, so changing
+   a variable is not taken into account. Cargo has an ad-hoc mechanism for doing
+   this in build scripts, but there's nothing to make this guaranteed correct
+   (i.e. that all variables accessed are tracked).
+2. There's no easy way to audit which environment variables a crate accesses.
+   This not only exacerbates the problem above, but it also means that
+   potentially sensitive information in an environment variable can be
+   incorporated into the compiled code.
+3. There's no way to override variables if they're needed by the build process
+   itself. For example, the `PATH` variable likely needs to be set so that the
+   compiler can execute various sub-processes, but there's no way to override
+   this so that `env!("PATH")` returns something else. This would be necessary
+   where the compilation environment differs from the deployment environment
+   (such as when cross-compiling).
+
+This RFC proposes a way to precisely control the environment visible to the
+compile-time macros, while defaulting to the current behaviour of making the
+entire environment available.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Rust implements the `env!()` and `option_env!()` macros to access a logical
+compile time environment at compilation time. By default, this logical
+environment is initialized from, and is identical to, the inherited process
+environment of `rustc` itself. (The only exception being environment variables
+whos names or values are not valid utf8.)
+
+There are several command-line options to control this environment and change it
+from the default:
+- `--env-clear` - remove all process environment variables from the logical
+  environment, leaving it completely empty.
+- `--env-remove VARIABLE` - Remove a specific variable from the logical
+  environment. This is an exact match, and it does nothing if that variable is
+  not set.
+- `--env-pass VARIABLE` - Pass a variable from the process environment to the
+  logical one, even if it had previously been removed. This lets specific
+  variables to be allow-listed without having to explicitly set their value. The
+  variable is ignored if it is not set or not utf8 encoded.
+- `--env-set VARIABLE=VALUE` - Set a variable in the logical environment. This will
+  either create a new variable, or override an existing value.
+
+The options are processed in the order listed above (ie, clear, then remove,
+then pass, then set). `--env-clear`, `--env-remove` and `--env-pass` are
```

Comment:
I think it would be better if these were always processed in the order they were passed. For example:

```sh
FOO=global rustc ... --env-set=FOO=override --env-remove=FOO --env-pass=FOO
```

My expectation is that `env!("FOO")` here should be `global`, but I think as described it would be `override`. Similarly for clear, it looks like `env!("FOO")` should fail with the below but I think it would be `bar`:

```sh
rustc ... --env-set=FOO=bar --env-clear
```

Reasoning:

1. Much easier to understand, you just read `--env-*` arguments left to right
2. More composable for build scripts since you can override your earlier settings. Roughly:
    ```sh
    # default args
    args="--env-clear --env-set=PATH=/bin --env-set=DYLIB=true"
    # a long build script...
    [ "$use_static_linking" = "true" ] && args="$args --env-remove=DYLIB"
    [ "$cross_compiling" = "false" ] && args="$args --env-pass=PATH"
    # ...
    ```

3. Slightly more straightforward implementation - just a `BTreeMap` with the system env, then each `--env-*` arg in order is an insert/remove/clear operation. No ordering of arguments needed

In any case, could you add some examples of CLI arguments and the resulting `env!`?

Reply:
Yeah I thought about that but it seemed like it would need some custom option parsing code to implement. And while I'm not adverse to the "read from left to right" argument, my pedantic side would actually prefer failing on multiple conflicting treatments of a specific variable rather than overriding. But I know pragmatically that the "keep adding args" pattern is very common for build rules, so maybe that's just too pedantic.

Reply:
I believe that the parser is already order-preserving, though can't say for certain.

If the decision is against making these composable, it could be reasonable to enforce the order of the args. That is, error if args aren't presented in `--env-clear`, `--env-remove` `--env-pass` `--env-set` order, even if it makes no difference, to enforce that they are always left to right readable.